File: submissions/exercise4/doron/grammar.py
#!/usr/bin/python


import tokengenerator


grammarlist = {}

class Grammar:
    
    def __init__(self, name):
        self.name = name
        self.grammarlist = []
        
class GrammarContent:
    
    def __init__(self,token,type):
        self.token = token
        self.type = type

def addGrammars():
    start = Grammar("START")
    start.grammarlist.append([GrammarContent("function", 1), GrammarContent("identifier", 3), GrammarContent("(", 1), GrammarContent("FUNCTION_ARGS", 2),GrammarContent(")", 1), GrammarContent("{", 1), GrammarContent("EXPR",2), GrammarContent("}", 1)])
    
    grammarlist["START"] = start
    
    function_args = Grammar("FUNCTION_ARGS")
    function_args.grammarlist.append([GrammarContent("identifier",3), GrammarContent("FUNCTION_CONT", 2)])
    function_args.grammarlist.append([GrammarContent("eps",4)])
    
    grammarlist["FUNCTION_ARGS"] = function_args
    
    function_cont = Grammar("FUNCTION_CONT")
    function_cont.grammarlist.append([GrammarContent(",", 1), GrammarContent("identifier",3)])
    function_cont.grammarlist.append([GrammarContent("eps",4)])
    
    grammarlist["FUNCTION_CONT"] = function_cont
    
    expr = Grammar("EXPR")
    expr.grammarlist.append([GrammarContent("VARDEC",2), GrammarContent("EXPR_CONT",2)])
    expr.grammarlist.append([GrammarContent("INPUT",2), GrammarContent("EXPR_CONT",2)])
    expr.grammarlist.append([GrammarContent("OUTPUT",2), GrammarContent("EXPR_CONT",2)])
    expr.grammarlist.append([GrammarContent("IF_ELSE",2), GrammarContent("EXPR_CONT",2)])
    expr.grammarlist.append([GrammarContent("WHILE_LOOP",2), GrammarContent("EXPR_CONT",2)])
    
    
    grammarlist["EXPR"] = expr
    
    expr_cont = Grammar("EXPR_CONT")
    expr_cont.grammarlist.append([GrammarContent("VARDEC",2), GrammarContent("EXPR_CONT",2)])
    expr_cont.grammarlist.append([GrammarContent("INPUT",2), GrammarContent("EXPR_CONT",2)])
    expr_cont.grammarlist.append([GrammarContent("OUTPUT",2), GrammarContent("EXPR_CONT",2)])
    expr_cont.grammarlist.append([GrammarContent("IF_ELSE",2), GrammarContent("EXPR_CONT",2)])
    expr_cont.grammarlist.append([GrammarContent("WHILE_LOOP",2), GrammarContent("EXPR_CONT",2)])
                                 
    expr_cont.grammarlist.append([GrammarContent("eps",4)])                                 

    grammarlist["EXPR_CONT"] = expr_cont
    
    vardec = Grammar("VARDEC")
    vardec.grammarlist.append([GrammarContent("identifier",3), GrammarContent("VARDEC_CONT",2), ])
    
    grammarlist["VARDEC"] = vardec
    
    vardec_cont = Grammar("VARDEC_CONT")
    vardec_cont.grammarlist.append([GrammarContent("EQUATION",2)])
    vardec_cont.grammarlist.append([GrammarContent("eps",4)])
    
    grammarlist["VARDEC_CONT"] = vardec_cont
     
    equation = Grammar("EQUATION")
    equation.grammarlist.append([GrammarContent("=",1), GrammarContent("VALID_EQUATORS",2)])
    
    
    grammarlist["EQUATION"] = equation
    
    valid_equators = Grammar("VALID_EQUATORS")
    
    valid_equators.grammarlist.append([GrammarContent("MATH_EXPR",2)])
    valid_equators.grammarlist.append([GrammarContent("string",3)])
    valid_equators.grammarlist.append([GrammarContent("integer",3)])
    valid_equators.grammarlist.append([GrammarContent("double",3)])
    valid_equators.grammarlist.append([GrammarContent("identifier",3)])
    
    
    grammarlist["VALID_EQUATORS"] = valid_equators
    
    input_ = Grammar("INPUT")
    input_.grammarlist.append([GrammarContent("get",1), GrammarContent("string",3)])
    input_.grammarlist.append([GrammarContent("get",1), GrammarContent("identifier",3)])
    
    grammarlist["INPUT"] = input_
    
    output = Grammar("OUTPUT")
    output.grammarlist.append([GrammarContent("print",1),GrammarContent("VALID_EQUATORS",2)])
    
    grammarlist["OUTPUT"] = output
    
    math_expr = Grammar("MATH_EXPR")
    
    math_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("math_op",3), GrammarContent("integer",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("math_op",3), GrammarContent("double",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("math_op",3), GrammarContent("identifier",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("math_op",3), GrammarContent("double",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("math_op",3), GrammarContent("identifier",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("math_op",3), GrammarContent("identifier",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("math_op",3), GrammarContent("integer",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    math_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("math_op",3), GrammarContent("integer",3), GrammarContent("MATH_EXPR_CONT",2)])
     
    math_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("math_op",3), GrammarContent("double",3), GrammarContent("MATH_EXPR_CONT",2)])
    
    
    
    grammarlist["MATH_EXPR"] = math_expr
    
    
    
    
    math_expr_cont = Grammar("MATH_EXPR_CONT")
    math_expr_cont.grammarlist.append([GrammarContent("math_op",3),GrammarContent("integer",3), GrammarContent("MATH_EXPR_CONT",2)])
    math_expr_cont.grammarlist.append([GrammarContent("math_op",3),GrammarContent("double",3), GrammarContent("MATH_EXPR_CONT",2)])
    math_expr_cont.grammarlist.append([GrammarContent("math_op",3),GrammarContent("identifier",3), GrammarContent("MATH_EXPR_CONT",2)])
    math_expr_cont.grammarlist.append([GrammarContent("eps",4)])
    
    grammarlist["MATH_EXPR_CONT"] = math_expr_cont
    
    
    if_else = Grammar("IF_ELSE")
    if_else.grammarlist.append([GrammarContent("if",1), GrammarContent("(",1), GrammarContent("COMP_EXPR",2), GrammarContent(")",1), GrammarContent("{",1), GrammarContent("EXPR",2), GrammarContent("}",1), GrammarContent("else",1), GrammarContent("{",1), GrammarContent("EXPR",2), GrammarContent("}",1)])
    if_else.grammarlist.append([GrammarContent("if",1), GrammarContent("(",1), GrammarContent("COMP_EXPR",2), GrammarContent(")",1), GrammarContent("{",1), GrammarContent("EXPR",2), GrammarContent("}",1)])
    
    grammarlist["IF_ELSE"] = if_else
    
    while_loop = Grammar("WHILE_LOOP")
    
    while_loop.grammarlist.append([GrammarContent("while",1), GrammarContent("(",1), GrammarContent("COMP_EXPR",2), GrammarContent(")",1), GrammarContent("{",1), GrammarContent("EXPR",2), GrammarContent("}",1)])
    
    grammarlist["WHILE_LOOP"] = while_loop
    
    
    comp_expr = Grammar("COMP_EXPR")
    comp_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("comp_op",3), GrammarContent("identifier",3)])
    comp_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("comp_op",3), GrammarContent("integer",3)])
    comp_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("comp_op",3), GrammarContent("double",3)])
    comp_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("comp_op",3), GrammarContent("integer",3)])
    comp_expr.grammarlist.append([GrammarContent("identifier",3), GrammarContent("comp_op",3), GrammarContent("double",3)])
    comp_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("comp_op",3), GrammarContent("identifier",3)])
    comp_expr.grammarlist.append([GrammarContent("integer",3), GrammarContent("comp_op",3), GrammarContent("double",3)])
    comp_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("comp_op",3), GrammarContent("identifier",3)])
    comp_expr.grammarlist.append([GrammarContent("double",3), GrammarContent("comp_op",3), GrammarContent("integer",3)])
    
    grammarlist["COMP_EXPR"] = comp_expr
    
    #X.grammarlist.append([])
    #GrammarContent("#",X)
    


class GrammarParsed:
    
    def __init__(self,token,grammar,type):
        self.token = token
        self.grammar = grammar
        self.type = type
class ReturnCatcher:
    
    def __init__(self):
        self.fail = False
        self.grammar = []
        self.error_messages = []
        self.index = 0;



def checkGrammar(grammarinp, index):
    
    
    
    s = grammarlist[grammarinp]
    ret = ReturnCatcher()
    fail = False
    tempindex = index
    
    for x in s.grammarlist:
        ret = ReturnCatcher()
        fail = False
        index = tempindex
        for y in x:
            #print(y.token)
            temp = tokengenerator.tokenlist[index]
            #print(str(index)+" "+s.name+" check"+temp.token+" "+ str(temp.value) +" with"+y.token)
            if(y.type == 1):
                if y.token.strip() is temp.token.strip():
                    ret.grammar.append(GrammarParsed(temp.token.strip(), s.name, "None" ))
                    index = index + 1
                else:
                    fail = True
                    
                    ret.error_messages.append("SYNTAX ERROR: EXPECTED"+y.token+" GOT " + temp.token)
                    
                    break
            elif(y.type == 2):
                temp_1 = checkGrammar(y.token, index)
                
                if temp_1.fail == True:
                    for errors in temp_1.error_messages:
                        ret.error_messages.append(errors)
                    fail = True
                    break
                else:
                    for gram in temp_1.grammar:
                        ret.grammar.append(gram)
                    index = temp_1.index
            elif(y.type == 3):
                if y.token.strip() is temp.token.strip():
                    ret.grammar.append(GrammarParsed(temp.value.strip(), s.name, temp.token ))
                    index = index + 1
                else:
                    fail = True
                    
                    ret.error_messages.append("SYNTAX ERROR: EXPECTED"+y.token+" GOT " + temp.token)
                    break
       
        if(fail == True):
            #print("fail")
            ret.fail = True
            #ret.error_messages = []
            ret.grammar = []
            ret.index = tempindex
            
        else:
            
            ret.fail = False
            ret.index = index
            return ret
        
   # print("fail2")
    ret.fail = True
   # ret.error_messages = []
    ret.grammar = []
    ret.index = index
    return ret
    

File: submissions/exercise3/tagarda/src/Grammar.java
import java.util.*;
/*
    GRAMMAR class

*/
class Grammar {
    static LinkedHashMap<String, Grammar> grammarlist = new LinkedHashMap<String, Grammar>(); //list of all grammars
    


    String name;
    ArrayList<GrammarConstruct []> g = new ArrayList<GrammarConstruct[]>(); //grammar parse tree database

    //grammar object definition
    public Grammar(String name){
        this.name = name;
        Grammar.grammarlist.put(this.name, this);
    }

    //a grammar contains multiple constructs can be terminal, nonterminal, or non-definites (String, Variables, Numbers)
    void addGrammar(GrammarConstruct[] constructs ){

        this.g.add(constructs);
    }

    //grammar definition (each grammar can be seen on grammar.txt)
     static void  constructGrammars() {
        //Grammar Construct Legend
        //1 - terminal
        //2 - nonterminal
        //3 - Strin
        Grammar FUNCTION_DEC = new Grammar("FUNCTION_DEC");
        FUNCTION_DEC.addGrammar(new GrammarConstruct[] {new GrammarConstruct("create", 1), new GrammarConstruct("VarID", 3),new GrammarConstruct("(", 1),new GrammarConstruct(")", 1), new GrammarConstruct("{", 1),new GrammarConstruct("EXPR", 2), new GrammarConstruct("}", 1)});

        Grammar EXPR = new Grammar("EXPR");
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("OUTPUT",2), new GrammarConstruct("EXPR_STMT",2)});
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VARASSIGN",2), new GrammarConstruct("EXPR_STMT",2)});
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("INPUT",2), new GrammarConstruct("EXPR_STMT",2)});
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VARDEC",2), new GrammarConstruct("EXPR_STMT",2)});
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("IF_STMT",2), new GrammarConstruct("EXPR_STMT",2)});
        EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("WHILE_LOOP",2), new GrammarConstruct("EXPR_STMT",2)});

        Grammar EXPR_STMT = new Grammar("EXPR_STMT");
        EXPR_STMT.addGrammar(new GrammarConstruct[]{new GrammarConstruct("EXPR", 2)});
        EXPR_STMT.addGrammar(new GrammarConstruct[]{new GrammarConstruct("eps",4)});

        Grammar OUTPUT = new Grammar("OUTPUT");
        OUTPUT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("meow", 1), new GrammarConstruct("String", 3), new GrammarConstruct(";", 1)});
        OUTPUT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("meow", 1), new GrammarConstruct("VarID", 3), new GrammarConstruct(";", 1)});
        OUTPUT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("meow", 1), new GrammarConstruct("NUMBER", 2), new GrammarConstruct(";", 1)});

        Grammar INPUT = new Grammar("INPUT");
        INPUT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("purr",1), new GrammarConstruct("VarID",3), new GrammarConstruct(";",1)});

        Grammar NUMBER = new Grammar("NUMBER");
        NUMBER.addGrammar(new GrammarConstruct[] { new GrammarConstruct("Integer",3)});
        NUMBER.addGrammar(new GrammarConstruct[] { new GrammarConstruct("Float",3)});

        Grammar MATH_EXPR = new Grammar("MATH_EXPR");
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("NUMBER",2), new GrammarConstruct("+",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("NUMBER",2), new GrammarConstruct("-",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("NUMBER",2), new GrammarConstruct("*",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("NUMBER",2), new GrammarConstruct("/",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("NUMBER",2), new GrammarConstruct("%",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("+",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("-",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("*",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("/",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("%",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});

        Grammar MATH_EXPR_EXTENSION = new Grammar("MATH_EXPR_EXTENSION");
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("+",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("-",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("*",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("/",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("%",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("+",1), new GrammarConstruct("VarID",3), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("-",1), new GrammarConstruct("VarID",3), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("*",1), new GrammarConstruct("VarID",3), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("/",1), new GrammarConstruct("VarID",3), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("%",1), new GrammarConstruct("VarID",3), new GrammarConstruct("MATH_EXPR_EXTENSION",2)});
        MATH_EXPR_EXTENSION.addGrammar(new GrammarConstruct[] {new GrammarConstruct("eps",4)});

        Grammar VARASSIGN = new Grammar("VARASSIGN");
        VARASSIGN.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("=", 1), new GrammarConstruct("NUMBER",2), new GrammarConstruct(";",1)});
        VARASSIGN.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("=", 1), new GrammarConstruct("VarID",3), new GrammarConstruct(";",1)});
        VARASSIGN.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("=", 1), new GrammarConstruct("String",3), new GrammarConstruct(";",1)});
        VARASSIGN.addGrammar(new GrammarConstruct[] {new GrammarConstruct("VarID",3), new GrammarConstruct("=", 1), new GrammarConstruct("MATH_EXPR",2), new GrammarConstruct(";",1)});

        Grammar VARDEC = new Grammar("VARDEC");
        VARDEC.addGrammar(new GrammarConstruct[] { new GrammarConstruct("make",1), new GrammarConstruct("VarID",3), new GrammarConstruct("VARDEC_CONT",2)});

        Grammar VARDEC_CONT = new Grammar("VARDEC_CONT");
        VARDEC_CONT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("=",1), new GrammarConstruct("NUMBER",2), new GrammarConstruct(";",1) });
        VARDEC_CONT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("=",1), new GrammarConstruct("VarID",3), new GrammarConstruct(";",1) });
        VARDEC_CONT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("=",1), new GrammarConstruct("String",3), new GrammarConstruct(";",1) });
        VARDEC_CONT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("=",1), new GrammarConstruct("MATH_EXPR",2), new GrammarConstruct(";",1) });
        VARDEC_CONT.addGrammar(new GrammarConstruct[] {new GrammarConstruct(";",1)});

        Grammar IF_STMT = new Grammar("IF_STMT");
        IF_STMT.addGrammar(new GrammarConstruct[] {new GrammarConstruct("meowif",1),new GrammarConstruct("(",1), new GrammarConstruct("COMP_EXPR",2), new GrammarConstruct(")",1), new GrammarConstruct("{",1), new GrammarConstruct("EXPR",2), new GrammarConstruct("}",1), new GrammarConstruct("ELSE_STMT",2)});

        Grammar ELSE_STMT = new Grammar("ELSE_STMT");
        ELSE_STMT.addGrammar(new GrammarConstruct[]{new GrammarConstruct("meowelse",1), new GrammarConstruct("{",1), new GrammarConstruct("EXPR",2), new GrammarConstruct("}",1)});
        ELSE_STMT.addGrammar(new GrammarConstruct[]{new GrammarConstruct("eps",4)});

        Grammar COMP_EXPR = new Grammar("COMP_EXPR");
        COMP_EXPR.addGrammar(new GrammarConstruct[]{ new GrammarConstruct("NUMBER",2), new GrammarConstruct("COMPARATORS",2), new GrammarConstruct("NUMBER",2) });
        COMP_EXPR.addGrammar(new GrammarConstruct[]{ new GrammarConstruct("NUMBER",2), new GrammarConstruct("COMPARATORS",2), new GrammarConstruct("VarID",3) });
        COMP_EXPR.addGrammar(new GrammarConstruct[]{ new GrammarConstruct("VarID",3), new GrammarConstruct("COMPARATORS",2), new GrammarConstruct("NUMBER",2) });
        COMP_EXPR.addGrammar(new GrammarConstruct[]{ new GrammarConstruct("VarID",3), new GrammarConstruct("COMPARATORS",2), new GrammarConstruct("VarID",3) });

        Grammar COMPARATORS = new Grammar("COMPARATORS");
        COMPARATORS.addGrammar(new GrammarConstruct[] { new GrammarConstruct(">",1), new GrammarConstruct("COMPARATORS2",2)});
        COMPARATORS.addGrammar(new GrammarConstruct[] { new GrammarConstruct("<",1), new GrammarConstruct("COMPARATORS2",2)});
        COMPARATORS.addGrammar(new GrammarConstruct[] { new GrammarConstruct("=",1), new GrammarConstruct("COMPARATORS2",2)});
        COMPARATORS.addGrammar(new GrammarConstruct[] { new GrammarConstruct("!",1), new GrammarConstruct("COMPARATORS2",2)});

        Grammar COMPARATORS2 = new Grammar("COMPARATORS2");
        COMPARATORS2.addGrammar(new GrammarConstruct[] { new GrammarConstruct("=",1)});
        COMPARATORS2.addGrammar(new GrammarConstruct[] { new GrammarConstruct("eps",4)});

        Grammar WHILE_LOOP = new Grammar("WHILE_LOOP");
        WHILE_LOOP.addGrammar(new GrammarConstruct[] { new GrammarConstruct("rollwhile", 1), new GrammarConstruct("(",1), new GrammarConstruct("COMP_EXPR",2), new GrammarConstruct(")",1), new GrammarConstruct("{",1), new GrammarConstruct("EXPR",2), new GrammarConstruct("}",1)});
    }

    //a recursive function that checks the grammar of the source code
    static MessageHandler CheckGrammar(Grammar start, int tokenindex,int tabcount) {
        boolean fail = false;
        int startindex = tokenindex;
        MessageHandler mh = new MessageHandler();
        MessageHandler mhtemp;
        String tc = "";

        //System.out.println(start.name+" length is:"+start.g+" "+tokenindex);
        //starting from start grammar, traverse and compare with the source code
        for(GrammarConstruct[] gram: start.g){

            fail = false;
            String grammar = "";
            mh.error_messages = new ArrayList<String>();
            

            //traverse a grammar definition
            //each gram is an array that contains something like this
            // { [meow,1], [VarID, 3], [;,1]}
            //which can be translated to
            // PRINT -> meow VARID ;

            for(int i=0; i<gram.length; i++){

                try{
                    Tokens temptoken = Tokens.tokenlist.get(tokenindex);
                    //if the current construct is a terminal
                    if(gram[i].type == 1){

                        if(gram[i].construct.trim().equals(temptoken.token.trim())){
                          

                            grammar = grammar + tc+temptoken.token.trim()+"\n";

                        }else{
                            fail = true;
                            mh.error_messages.add(tc+"SYNTAX ERROR: EXPECTED "+gram[i].construct+" GOT "+temptoken.token+"\n");
                        //    System.out.println(tc+"SYNTAX ERROR: EXPECTED "+gram[i].construct+" GOT "+temptoken.token+"\n");
                            break;
                        }
                        tokenindex++;
                        //if the current construct is a nonterminal
                    }else if(gram[i].type == 2){
                        //recursively call the nonterminal grammar definition
                        mhtemp = CheckGrammar(Grammar.grammarlist.get(gram[i].construct), tokenindex,tabcount++);
                        //check if it returns a valid grammar
                        if(mhtemp.isFailed == true){
                            fail = true;
                            for(String k:mhtemp.error_messages){
                                mh.error_messages.add(k);
                            }
                            break;
                        }else {
                            tokenindex = mhtemp.tokenindex;
                            grammar = grammar + mhtemp.grammar;
                            for(String k:mhtemp.error_messages){
                                mh.error_messages.add(k);
                            }
                            
                        }
                        //check for string, variables, etc.
                    }else if(gram[i].type == 3){

                        if(gram[i].construct.trim().equals(temptoken.value.trim())){
                            
                            grammar = grammar +tc+temptoken.token.trim();
                            //System.out.println(tc+temptoken.token.trim());
                        }else{
                            fail = true;

                            mh.error_messages.add(tc+"SYNTAX ERROR: EXPECTED "+gram[i].construct+" GOT "+temptoken.token+"\n");
                        //    System.out.println(tc+"SYNTAX ERROR: EXPECTED "+gram[i].construct+" GOT "+temptoken.token+"\n");
                            break;
                        }
                        tokenindex++;
                    }else if(gram[i].type == 4){
                        mh.isFailed = false;
                        mh.tokenindex = tokenindex;
                        mh.grammar = grammar;
                        return mh;
                    }
                }catch(Exception e){

                    mh.error_messages.add("Missing "+gram[i].construct+"\n");
                    mh.isFailed = true;
                    mh.tokenindex = tokenindex;
                    mh.grammar = grammar;
                    return mh;
                }
            }

            //reset the search if the current grammar definition fails (check other grammars possible)
            if (fail == true){
                //System.out.println("Failed.");
                tokenindex = startindex;
            }else {
                //otherwise return a valid grammar definition
             mh.isFailed = false;
             mh.tokenindex = tokenindex;
             mh.grammar = grammar;

             return mh;
            }
        }

        //failed grammar (no valid definition)
        mh.isFailed = true;
        mh.tokenindex = tokenindex;
        return mh;
    }
}
